*** macros-derived-orig.scm	Mon Sep 18 20:35:48 2006
--- macros-derived.scm	Mon Sep 18 20:39:28 2006
***************
*** 63,68 ****
--- 63,72 ----
     ;; R6RS additions:
     
     unspecified
+ 
+    ;; Additions for records:
+ 
+    error
     
     ;; Additions defined in the core expander:
     
***************
*** 132,137 ****
--- 136,145 ----
         ;; R6RS additions:
         
         unspecified
+ 
+        ;; Additions for records:
+ 
+        error
         
         ;; Additions defined in the core expander:
         
***************
*** 592,598 ****
--- 600,611 ----
     ;; R6RS additions:
     
     unspecified
+ 
+    ;; Additions for records:
     
+    error (rename (error contract-violation))
+    forall exists find let-values
+  
     ;; Additions defined in the core expander:
     
     begin if set! lambda quote
***************
*** 621,626 ****
--- 634,664 ----
            (for (core identifier-syntax) run expand)      
            (for (core quasisyntax)       run expand)        
            (for (core quasiquote)        run expand))    
+ 
+   ;; Additions for records:
+ 
+   (define-syntax let-values
+     (syntax-rules ()
+       [(_ ([(id ...) expr]) body ...)
+        (call-with-values (lambda () expr)
+ 	 (lambda (id ...) body ...))]))
+   (define (find f l)
+     (cond
+      [(null? l) #f]
+      [(f (car l)) (car l)]
+      [else (find f (cdr l))]))
+   (define (forall f l l2)
+     (cond
+      [(null? l) #t]
+      [(null? (cdr l)) (f (car l) (car l2))]
+      [(f (car l) (car l2)) (forall f (cdr l) (cdr l2))]
+      [else #f]))
+   (define (exists f l)
+     (cond
+      [(null? l) #f]
+      [(null? (cdr l)) (f (car l))]
+      [(f (car l)) => (lambda (v) v)]
+      [else (exists f (cdr l))]))
    
    ) ;; r6rs
  
